### ISSUE TITLE
[v3] Configuration changes/renames/deprecations/etc

### ISSUE THREAD

**Issue by Maikuolan** [CIDRAM/CIDRAM#139]
Creating this issue to track ideas/suggestions/etc in regards to configuration changes proposed since after the first v2 release, which aren't backwards-compatible, and therefore must be postponed until work begins on v3.

Creating such an issue is useful as a means to ensure that we don't forget any such proposed changes. However, since the first v2 release was quite recent, and a reasonable amount of time should be provided between major releases in order to not pose a major headache for end-users, we don't yet have any particular timeframe in mind for when to begin work on v3. This means that this particular issue most likely won't see much activity for quite a while yet, and should remain open during that time.

---

First (and only current) proposal:

*In reference to: https://github.com/CIDRAM/CIDRAM/issues/136#issuecomment-511185560*

*Excerpt:*
> > forbid_on_block option should be named HTTP_RESPONSE_HEADER_code imho, as it is a bit confusing and ambiguous as it is now.
>
> Fair call. :-)
>
> The main reason for the current name, is because of how that configuration directive worked back in the early days of CIDRAM, when it was first implemented. Back when it was first implemented, it acted just as a simple on/off switch (true/false), for whether to send a 403 header ("forbidden") when a request is blocked (i.e., *should CIDRAM "forbid on block"?*), or whether to just leave it be and allow the standard 200 OK header to be sent instead. The various other options (410, 418, 451, 503) were added later, during various stages of development.
>
> I agree though, that something like `http_response_header` or similar would make more sense as a name, considering how it works at this point. Maybe renaming this configuration directive is something we should earmark for v3.

**Comment by Maikuolan** [comment]
Possible (backwards-incompatible) improvement to the logic used for configuration directives (inspired by the PR that previously referenced this issue): Consistently use a negative-negative/positive-positive logic for configuration directives (i.e., change/remove/replace any existing configuration directives that currently use a positive-negative/negative-positive logic). Configuration directives that would be affected by this proposal:
- `disable_frontend` (defaults to `true`). Could replace with `enable_frontend` (defaulting to `false`).
- `hide_version` (defaults to `false`). Could replace with `show_version` (defaulting to `true`).
- The various field omission configuration directives (`omit_ip`, `omit_hostname`, `omit_ua`). Thinking currently that I should deprecate all three, in favour of a singular field selection configuration directive (when I implemented channel switching, in order to be able to better support channel switching, I also implemented a new type of configuration directive, which wasn't previously supported by CIDRAM, "checkbox"; this is why the configuration page shows checkboxes for that particular configuration directive, but not for any of the other configuration directives; the field omission configuration directives were already implemented by that point, prior to this new configuration directive type being supported; thinking that if I replaced these with a singular field selection configuration directive, that would both satisfy the negative-negative/positive-positive proposal, as well as allow for better control for the end-user, to specify exactly which fields they want logged, without needing separate configuration directives for each of them).

**Comment by ZB-Block-JamesC** [comment]
> _Consistently use a negative-negative/positive-positive logic for configuration directives_

Have you ever watched any courtroom television shows, where a lawyer tries to paint a witness as uncreditable by asking an obviously false question? 
Lawyer, after several "yes" responses in a row: "Isn't it true that you were once a dog?" 
Witness: "Yes. I mean no! I was never a dog, why would you ask such a thing?"
Lawyer, smirking at the jury: "No further questions."

That's where you are with the negative-negative/positive-positive logic. The names of the options should reflect their **default setting as though you were verbally asking**; any change by the site admin would invert (or cancel out, or negate) the name of the option -- just as a verbal "no" would invert the response you expect. :)

**disable_frontend** with a default of **true** is fine as it is. In spoken English, a site admin might literally say "no, I don't want this disabled", which is the verbal equivalent of setting this to **false**. 
 
**hide_version** with a default of **false** is backward. Your suggestion, **show_version** with a default of **true**, is more consistent with spoken English: "No, I don't want this to show" = setting this to **false**. 

I realize that in writing, this doesn't line up with negative-negative/positive-positive ... but when site admins read the various configuration options and think "yes" or "no" to each setting, this makes sense. :)

**Comment by Maikuolan** [comment]
Fair call. :-)

In any case, we definitely want it to sound as natural and intuitive as possible (the point being, after all, to avoid mistakes or misunderstandings occurring due to counter-intuitiveness).

Plenty of time to figure it out though; not likely to start on this for quite a while yet still (at least, not until v2 is "mature", and not until we're ready to retire v1 into EoL status).

**Comment by Maikuolan** [comment]
*Partially* implemented (5e486bf5a0e94a6cce5c6524e4ff78ff252b0988).

*Changelog excerpt:*
```
- Configuration directives "hide_version", "empty_fields", "omit_ip",
  "omit_hostname", and "omit_ua" have been replaced by a new configuration
  directive, "fields". The new configuration directive is capable of
  providing the same functionality provided previously, as well as some
  other, new, related functionality.
- Configuration directive "error_log_stages" has been replaced by a new
  configuration directive, "stages". The new configuration directive is
  capable of providing the same functionality provided previously, as well
  as some other, new, related functionality.
```

New configuration directives provide checkboxes at the front-end, allowing users to explicitly specify *exactly* which fields to show/hide and so on.

Still have a few other configuration directives to be dealt with yet though (e.g., `disable_frontend`).

**Comment by Maikuolan** [comment]
*Changelog excerpt:*

```
- Configuration directives "block_attacks", "block_cloud", "block_bogons",
  "block_generic", "block_legal", "block_malware", "block_proxies", and
  "block_spam" have been replaced by a new configuration directive,
  "shorthand". The new configuration directive is capable of providing the
  same functionality provided previously, as well as some other, new,
  related functionality.
```

With that, everything described by this issue has been done. :-)

Marking as implemented and closing.
